corva/models/stream.py

```diff
@@ -57,20 +72,61 @@ def app_stream_id(self) -> int:
 
     @property
     def is_completed(self) -> bool:
+        """there can only be 1 completed record, always located at the end of the list"""
+
         if self.records:
             return self.records[-1].collection == 'wits.completed'
 
         return False
 
 
-class StreamEvent(ListEvent[StreamEventData]):
+class StreamEvent(BaseEvent, StreamEventData):
     @staticmethod
-    def from_raw_event(event: str, **kwargs) -> StreamEvent:
-        app_key: str = kwargs['app_key']
+    def from_raw_event(event: str, **kwargs) -> List[StreamEvent]:
+        app_key = kwargs['app_key']
+
+        events = pydantic.parse_raw_as(List[StreamEvent], event)  # type: List[StreamEvent]
+
+        for event in events:
+            event.app_key = app_key
+
+        return events
+
+    @staticmethod
+    def filter(
+         event: StreamEvent, by_timestamp: bool, by_depth: bool, last_timestamp: int, last_depth: float
+    ) -> StreamEvent:
+        records = event.records
+
+        if event.is_completed:
+            # there can only be 1 completed record, always located at the end of the list
+            records = records[:-1]  # remove "completed" record
+
+        new_records = []
```

Comment:
I think this should be a generator

Reply:
Generators are better to use in simple cases. In this case it'd be complex smth like
```python
        new_records = [
            record
            for record in records
            if (
                 (by_timestamp and record.timestamp > last_timestamp)
                 or
                 (by_depth and record.measured_depth > last_depth)
                 or (not by_timestamp and not by_depth)
            )
        ]
```

Reply:
I only mean something like this, which could be called from `filter()`.

```suggestion
        def process_records(self, records):
            for record in records:
                if by_timestamp and record <= last_timestamp:
                    continue
                
                if by_depth and record.measured_depth <= last_depth:
                    continue
                    
                yield record
```

Reply:
cool, I guess this will also work faster